src/build.c

```diff
@@ -712,17 +709,16 @@ static gchar *build_replace_placeholder(const GeanyDocument *doc, const gchar *s
 	else if (strstr(stack->str, "%p"))
 	{   /* fall back to %d */
 		ui_set_statusbar(FALSE, _("failed to substitute %%p, no project active"));
-		if (doc != NULL && filename != NULL)
-			replacement = g_path_get_dirname(filename);
+		if (doc != NULL && doc->file_name != NULL)
+			replacement = g_path_get_dirname(doc->file_name);
 	}
 
 	utils_string_replace_all(stack, "%p", replacement);
 	g_free(replacement);
 
-	ret_str = utils_get_utf8_from_locale(stack->str);
+	ret_str = stack->str;
 	g_free(executable);
-	g_free(filename);
-	g_string_free(stack, TRUE);
+	g_string_free(stack, FALSE);
```

Comment:
This should probably rather be rewritten as `return g_string_free(stack, FALSE);`